README.md
# perl-physics-ballistics
Internal, External, Terminal Ballistics Formulae, in Perl

Documentation follows for each of the three modules (Internal, External, Terminal), all appended together.  I'll figure out a nicer way to present this later.

# NAME

Physics::Ballistics::Internal -- Various internal ballistics formulae.

# ABSTRACT

Internal ballistics is the study of what happens while a projectile is being
launched from a barrel, from the time the propellant ignites to the moment
after the projectile leaves the muzzle.

This module implements a variety of functions and mathematical formulae useful
in the analysis and prediction of internal ballistic effects.

It also branches out somewhat into the closely-related matters of bullet and
cartridge composition and characteristics.

# REGARDING BULLET DIAMETERS

Some of these functions require the diameter of a projectile as a parameter.
Please note that bullet diameters are usually different from the names of
their calibers.  NATO 5.56mm bullets are actually 5.70mm in diameter, while
Russian 5.45mm bullets are actually 5.62mm.  .308 caliber bullets really are
0.308 inches in diameter (7.82mm), but .22 Long Rifle bullets are 0.222
inches across.

Making assumptions can hurt!  It's better to look it up before plugging it in.

# ANNOTATIONS OF SOURCES

Regarding their source, these functions fall into three categories:  Some are
simple encodings of basic physics (like energy = 1/2 \* mass \* velocity\*\*2),
and these will not be cited.  Others are from published works, such as books
or trade journals, and these will be cited when possible.  A few are products
of my own efforts, and will be thus annotated.

# OOP INTERFACE

A more integrated, object-oriented interface for these functions is under
development.

# CONSTANTS

## %CASE\_CAPACITY\_GRAINS

This hash table maps the names of various cartridges to their volumes, in grains of water.

It is based on the table at: http://kwk.us/cases.html

Powder capacities will vary considerably depending on powder type, but multiplying water capacity by 0.85 comes pretty close for most powder types.

# FUNCTIONS

## cartridge\_capacity (bullet\_diameter\_mm, base\_diameter\_mm, case\_len\_mm, psi, \[want\_powder\_bool\])

This function estimates the internal capacity (volume) of a rifle cartridge, assuming typical brass material construction.

It is not very accurate (+/- 6%) and actual volumes will vary between different manufacturers of the same cartridge anyway.

Its main advantage over a Powley calculator is ease of use, as it requires fewer and easier-to-obtain parameters.

Unlike the Powley calculator, it is only valid for typically-tapered rifle casings for calibers in the range of about 5mm to 14mm.

This function is the original work of the author.

DISCLAIMER:  Do not use this as a guide for handloading!  Get a Speer Reloading Manual or similar.  If you must use this function, start at least 15% lower than the function indicates and work your way up slowly.  Author is not responsible for idiots blasting copper case-bits into their own faces.

> parameter: (float) bullet\_diameter\_mm is the width of the bullet (in mm)
>
> parameter: (float) base\_diameter\_mm is the width of the case at its base, NOT its rim (in mm)
>
> parameter: (float) case\_len\_mm is the overall length of the case, including rim and neck (in mm)
>
> parameter: (float) psi is the peak pressure tolerated by the cartridge (in psi)
>
> parameter: (boolean) OPTIONAL: set want\_powder\_bool to a True value to approximate powder capacity instead of water capacity.  This is intrinsically inaccurate, since different powders have different weights, but it does okay for typical powders.
>
> returns: (int) cartridge capacity (in grains)

## empty\_brass (bullet\_diameter\_mm, base\_diameter\_mm, case\_len\_mm, psi)

This function estimates the weight of an empty rifle cartridge, assuming typical brass material construction.  This is just the weight of the case metal, without primer, bullet, or powder.

It is moderately accurate, with the occasional winger, but actual weights will vary between different manufacturers of the same cartridge anyway.

Its main advantage over a Powley calculator is ease of use, as it requires fewer and easier-to-obtain parameters.

Unlike the Powley calculator, it is only valid for reasonably-tapered rifle casings for calibers in the range of about 5mm to 14mm.

This function is the original work of the author.

Compare to known dimensions and weights:

    5.56x45mm:  95 gr, empty_brass predicts  95 = 1.000 actual, 0.0% error
    7.62x39mm: 100 gr, empty_brass predicts 105 = 1.050 actual, 5.0% error
    6.8mm SPC: 107 gr, empty_brass predicts 100 = 0.935 actual, 6.5% error
    7.62x51mm: 182 gr, empty_brass predicts 181 = 0.995 actual, 0.5% error
    12.7x99mm: 847 gr, empty_brass predicts 830 = 0.980 actual, 2.0% error

> parameter: (float) bullet\_diameter\_mm is the width of the bullet (in mm)
>
> parameter: (float) base\_diameter\_mm is the width of the case at its base, NOT its rim (in mm)
>
> parameter: (float) case\_len\_mm is the overall length of the case, including rim and neck (in mm)
>
> parameter: (float) psi is the peak pressure tolerated by the cartridge (in psi)
>
> parameter: (boolean) OPTIONAL: set want\_powder\_bool to a True value to approximate powder capacity instead of water capacity.  This is intrinsically inaccurate, since different powders have different weights, but it does okay for typical powders.
>
> returns: (int) cartridge weight (in grains)

## gunfire (psi, bullet\_diameter\_mm, barrel\_length\_inches, cartridge\_diameter\_mm, cartridge\_length\_mm, bullet\_mass\_gr)

The gunfire function attempts to approximate the performance of a cartridge/bullet/barrel combination in terms of muzzle velocity (and some related attributes).

It does a tolerable job, despite grossly oversimplifying the problem.  Its principle charms are that it is easy to use, and there isn't much else available for solving this kind of problem.  Improving this function is on my to-do list.

This function is the original work of the author.

Compare to known cartridge/bullet/barrel performance:

    270 Winchester Short Magnum, 140gr, 24" barrel:
        actual velocity:     3250 ft/s
        gunfire() predicts:  3489 ft/s (7% error)

    .300 Lapua Magnum, 185gr, 27" barrel:
        actual velocity:     3300 ft/s
        gunfire() predicts:  3527 ft/s (6% error)

    .25-06 Remington, 120gr, 24" barrel:
        actual velocity:     2990 ft/s
        gunfire() predicts:  3056 ft/s (2% error)

    .308 Winchester, 150gr, 24" barrel:
        actual velocity:     2820 ft/s
        gunfire() predicts:  2840 ft/s (1% error)

    12.7x99mm NATO, 655gr, 45" barrel:
        actual velocity:     3029 ft/s
        gunfire() predicts:  3033 ft/s (0.1% error)

    .223 Remington, 55gr, 24" barrel:
        actual velocity:     3240 ft/s
        gunfire() predicts:  3170 ft/s (2% error)

    .30-06, 180gr, 24" barrel:
        actual velocity:     2700 ft/s
        gunfire() predicts:  2580 ft/s (5% error)

    .375 Ruger, 300gr, 23" barrel:
        actual velocity:     2660 ft/s
        gunfire() predicts:  2431 ft/s (9% error)

In particular, take the "r,m" output field with a huge grain of salt.  For a more accurate number, use Physics::Ballistics::External::flight\_simulator().  The only advantage of "r,m" is that it is much much faster to derive (25 microseconds for gunfire(), vs an eighth of a second for flight\_simulator() on my hardware).

> parameter: (float) peak chamber pressure (in psi, NOT cup!)
>
> parameter: (float) bullet diameter (in mm)
>
> parameter: (float) barrel length (in inches)
>
> parameter: (float) cartridge base diameter (in mm)
>
> parameter: (float) cartridge overall length (in mm)
>
> parameter: (float) bullet mass (in grains)
>
> returns: a reference to a hash, with the following fields:
>
>     N*m: (int) muzzle energy (in joules)
>     f/s: (float) muzzle velocity (in feet per second)
>     m/s: (float) muzzle velocity (in meters per second)
>     r,m: (int) approx range achieved when fired at a 45 degree angle (in meters)
>     tm:  (float) time elapsed from ignition to bullet's egress from barrel (in seconds)

## ogival\_volume (length\_mm, radius\_mm, \[C,\] \[granularity\_mm\])

This function calculates the volume of a Haak-series ogival nose shape, as often used for areodynamically
streamlined projectiles.  It is useful (for instance) for determining the mass of a nose, when nose 
composition (and therefore density) is known.

qv: http://en.wikipedia.org/wiki/Nose\_cone\_design#Haack\_series

Quoting from that article:

    While the series is a continuous set of shapes determined by the value of
    C in the equations below, two values of C have particular significance:
    when C = 0, the notation LD signifies minimum drag for the given length
    and diameter, and when C = 1/3, LV indicates minimum drag for a given
    length and volume.  The Haack series nose cones are not perfectly tangent
    to the body at their base except for case where C = 2/3.  However, the
    discontinuity is usually so slight as to be imperceptible.  For C > 2/3,
    Haack nose cones bulge to a maximum diameter greater than the base diameter.
    Haack nose tips do not come to a sharp point, but are slightly rounded.

> parameter: (float) the length of the ogive, from base to tip (in mm)
>
> parameter: (float) the radius of the cross-section of the ogive (in mm)
>
> parameter: (float) OPTIONAL: the sharpness factor of the ogive, higher values providing a more fat, blunt nose shape (in range 0..2/3, default=2/3)
>
> parameter: (float) OPTIONAL: the granularity at which the volume will be calculated, lower values providing more accuracy but requiring more processing time (in mm, default=1/10000, provides < 0.1% error)
>
> returns: (float) volume (in cc)

## powley (bore\_diameter\_inches, case\_base\_diameter\_inches, case\_length\_inches, barrel\_1\_length\_inches, barrel\_2\_length\_inches)

This function implements Powley's formula for approximating the projectile velocity gained or lost from a change in barrel length.

Example of use:

    It is known that the muzzle velocity of a .223 Remington, 55gr bullet from a 24" barrel is 3240 ft/s.
    We want to know its muzzle velocity from a 16" barrel.

    powley (0.224, 0.378, 1.77, 24, 16) = 0.9205
    3240 ft/s * 0.9205 = 2982 ft/s

> parameter: (float) barrel's bore diameter (in inches)
>
> parameter: (float) cartridge's base case diameter (in inches)
>
> parameter: (float) cartridge's overall length (in inches)
>
> parameter: (float) the length of the barrel for which muzzle velocity is known (in inches)
>
> parameter: (float) the length of the barrel for which muzzle velocity is not known (in inches)
>
> returns: (float) the ratio of the muzzle velocities (unitless)

## cup2psi\_linear (cup\[, want\_range\[, fractional\_deviation\]\])

Approximates peak chamber pressure, in psi, given peak chamber CUP (copper crush test).  Since there is a degree of error present in both kinds of pressure tests, this will often disagree with published measurements.  To offset this, a range may be requested by passing a non-false second parameter.  This will cause three values to be returned:  A low-end psi estimate, the median psi estimate (which is the same as the value returned when called without a want\_range parameter), and a high-end psi estimate.  The degree of variation may be adjusted by passing a value between 0 and 1 as the third argument (default is 0.05).

Based on linear formula from Denton Bramwell's \_Correlating PSI and CUP\_, with curve-fitting enhancements by module author.

## cup2psi (cup\[, want\_range\[, fractional\_deviation\]\])

Approximates peak chamber pressure, in psi, given peak chamber CUP (copper crush test).  Since there is a degree of error present in both kinds of pressure tests, this will often disagree with published measurements.  To offset this, a range may be requested by passing a non-false second parameter.  This will cause three values to be returned:  A low-end psi estimate, the median psi estimate (which is the same as the value returned when called without a want\_range parameter), and a high-end psi estimate.  The degree of variation may be adjusted by passing a value between 0 and 1 as the third argument (default is 0.04).

Based on exponential formula from http://kwk.us/pressures.html, with enhancements by module author.

## recoil\_mbt (gun\_mass\_kg, projectile\_mass\_kg, projectile\_velocity\_mps, \[gas\_mass\_kg,\] \[gas\_velocity\_mps,\] \[recoil\_distance\_cm,\] \[english\_or\_metric\_str\])

Approximates the recoil force of a battletank's large-bore main gun (or any other large-bore, high-velocity gun).

Based on formula from Ogorkiewicz's \_Design and Development of Fighting Vehicles\_, page 58.

As a rule of thumb, the recoil force of an MBT-proportioned vehicle's main gun should not exceed twice the vehicle's mass.

If combustion gas mass and velocity are absent, they will be estimated from the projectile mass and velocity.

The gun mass includes all of the parts moving against the vechicle's recoil mechanism (principally, the barrel and breech).

> parameter: (float) gun mass (in kg)
>
> parameter: (float) projectile mass (in kg)
>
> parameter: (float) projectile muzzle velocity (in meters per second)
>
> parameter: (float) OPTIONAL: combustion gas mass, equal to the propellant mass, usually between one and one half the projectile mass (in kg)
>
> parameter: (float) OPTIONAL: combustion gas velocity (in meters per second, usually 1450).
>
> parameter: (float) OPTIONAL: recoil distance (in cm, default=20)
>
> returns: (float) recoil force exerted on the vehicle (in tonnes)

# TODO

The accuracy of these estimating functions can be improved, and I intend to improve them.

In particular, empty\_brass should be made to take a "parent case" option, because it tends to underestimate the weight of cartridges which are based on other cartridges which have been trimmed or necked down.


# NAME

Physics::Ballistics::External -- External ballistics formulae.

# ABSTRACT

External ballistics is the study of projectiles in flight, from the time they
leave the barrel (or hand, or trebuchet, or whatever), to the moment before
they strike their target.  This module implements mathematical formulae and
functions useful in the analysis and prediction of external ballistic behavior.

# ANNOTATIONS OF SOURCES

Regarding their source, these functions fall into three categories:  Some are
simple encodings of basic physics (like energy = 1/2 \* mass \* velocity\*\*2),
and these will not be cited.  Others are from published works, such as books
or trade journals, and these will be cited when possible.  A few are products
of my own efforts, and will be thus annotated.

# OOP INTERFACE

A more integrated, object-oriented interface for these functions is under
development.

# FUNCTIONS

## ebc (mass\_grains, diameter\_inches, \[shape,\] \[form\_factor\])

Attempts to predict the G1 ballistic coefficient of a projectile, based on its
mass, width, and shape.  Useful for predicting the ballistic behavior of
hypothetical or new, untested projectiles.  When compared against the known G1
BC's of well-understood projectiles, its predictions usually come within 5% of
actual.

The "shape" parameter indicates that the hypothetical projectile is closest in
its shape, composition, and quality of manufacture to the named entity.  Some
shapes are very general ("hollowpoint", "fmj") while others are very specific
to product families manufactured by particular companies ("scenar", "amax").

The "shape" parameter maps to a numerical form base (see table below), which
gets tweaked a little by other factors to derive a form factor.  The form
factor has an inverse impact on ballistic coefficient (higher form factor
means a lower coefficient).  In lieu of depending on the "shape" parameter, a
form factor may be specified as a parameter (mostly useful for debugging or
deriving new entries in the shape table).

When no shape or form factor are provided, the "default" shape is used, which
has a form base chosen to minimize the error produced by this function when
its output is compared to entries in http://www.frfrogspad.com/g1bclist.xls
not already covered in the shape table.

A hashfile including all of the information from g1bclist.xls with "shape"
entries matching the names from the shape table below may be found here:
http://ciar.org/ttk/mbt/guns/table.bc.hash

This function is the original work of the module's author.

> parameter: (float) mass of the projectile (in grains)
>
> parameter: (float) diameter of the projectile (in inches)
>
> parameter: (str) OPTIONAL: shape/composition of the projectile (see table below, default is "default")
>
> parameter: (float) OPTIONAL: custom form-factor of the projectile, unnecessary if "shape" is provided.
>
> returns: a list, containing the following values:
>
>     * The estimated G1 ballistic coefficient,
>     * The form factor (suitable for use as as form_factor parameter)
>     * The "very short factor" (1.0 for most well-proportioned bullets)
>     * The shape parameter used

The shape table currently contains the following entities, with the given form bases:

    '7n14'              => 111, # Very deep ogival shape and boat-tail with tight tolerances, used in military 7.62x54R specifically for sniping.
    'scenar'            => 124, # Scenar, by Lapua
    'scenar_s'          => 124, # Scenar Silver, by Lapua, appears ballistically indistinguishable from Scenar
    '7n1'               => 125, # Very deep ogival shape and boat-tail, used in military 7.62x54R.  qv http://7.62x54r.net/MosinID/MosinAmmo007.htm
    '7n6'               => 125, # Very deep ogival shape and boat-tail, used in military 5.45x39mm  qv http://7.62x54r.net/MosinID/MosinAmmo007.htm
    '7n6m'              => 125, # Synonym for 7N6
    'amax'              => 127, # A-Max, by Hornady
    'boat_tail_og'      => 128, # Catch-all for many boat-tailed projectiles with long, pointed ogival nose shapes
    'hollowpoint_ct'    => 131, # CT variation of hollowpoint "match" projectiles, by Nosler
    'bst'               => 136, # A type of flat-bottomed ogival, by Nosler
    'spire_point'       => 138, # A type of flat-bottomed ogival, by Speer
    'boat_tail_nosler'  => 138, # Nosler's line of boat-tails perform more poorly than others for some reason
    'boat_tail_ct'      => 139, # Nosler's CT variation of boat-tail
    'spitzer'           => 139, # Another flat-bottomed ogival, by Speer.  Good fit to many military bullets.
    'hollowpoint_match' => 140, # Catch-all for many hollowpoint "match" projectiles
    'accubond'          => 142, # An expansive line by Nosler offering controlled terminal expansion
    'interbond'         => 142, # A line of polymer-tipped bullets by Hornady optimized for terminal cohesion
    'vmax'              => 145, # A line of polymer-tipped bullets by Hornady optimized for terminal expansion / fragmentation
    'gold_match'        => 147, # A type of wadcutter, by Speer
    'grand_slam'        => 159, # Another wadcutter, by Speer, with less streamlined nose shape for higher overall mass
    'hollowpoint'       => 176, # Catch-all for many large-game hollowpoint projectiles
    'default'           => 179, # Catch-all for unknown/unspecified shapes; derived via best-fit to ballistic table
    'fmj'               => 187, # Catch-all for military full metal jacket with boat-tail and short nose
    'mag_tip'           => 190, # Another wadcutter, by Speer
    'afss'              => 202,
    'tsx_fb'            => 205, 
    'plinker'           => 205, # Typical of short, underweight, round-nosed, non-streamlined projectiles,
    'semispitzer'       => 224, # A foreshortened, flat-bottomed projectile, by Speer
    'round_nose'        => 228, # Catch-all for many flat-bottomed, hemispherical-nosed projectiles
    'varminter'         => 232, # Catch-all for many light hollowpoints with very large expanding cavities, for varminting
    'fmj_2'             => 268  # Woodleigh's FMJ projectiles, which are shape-optimized for travel in big game meat and bone, instead of air.

This hash table is exported as %Physics::Ballistics::External::Bullet\_Form\_Factors\_H,
so that users and modules may easily modify/add its content without resorting
to editing sources.

## flight\_simulator (drag\_function, ballistic\_coefficient, muzzle\_velocity\_fps, sight\_height\_inches, shot\_angle\_deg, \[bore\_to\_sight\_angle\_deg,\] zero\_range\_yards, wind\_speed\_fps, wind\_angle\_deg, \[max\_range\_yards\])

Attempts to predict the flight characteristics of a projectile in flight, providing a data point for every yard of progress it makes downrange.

This is a pure-perl port of the "SolveAll" function from GNU-Ballistics gebc-1.07 lib/ballistics/ballistics.cpp, and as slow as one might expect from a pure-perl port of an arithmetic-intensive algorithm.

On my system it takes about an eighth of a second to simulate a 1200-yard flight.  This may be supplemented at some point with an Inline::C equivalent.

Note that most manufacturers report G1 ballistic coefficients.  Using the wrong drag function for a given ballistic coefficient will produce ludicrously incorrect results.

To ascertain the correct bore elevation to hit a target at a specific distance, change the shot\_angle\_deg parameter on successive calls to flight\_simulator(), and converge on drop\_inches == 0.0 at the given range via binary search.  I should get around to providing a function for that at some point (GNU-Ballistics has such a function, I just didn't port it).

> parameter: (str) drag\_function is exactly one of: 'G1', 'G2', 'G5', 'G6', 'G7', 'G8'.
>
> parameter: (float) ballistic\_coefficient, qv: http://en.wikipedia.org/wiki/Ballistic\_coefficient
>
> parameter: (float) muzzle\_velocity\_fps is the velocity of the projectile at time=0 (feet per second)
>
> parameter: (float) sight\_height\_inches is the distance from the center of the sight to the center of the bore (inches)
>
> parameter: (float) shot\_angle\_deg is the bore elevation (degrees, 0 = horizontal, 90 = vertical)
>
> parameter: (float) OPTIONAL: bore\_to\_sight\_angle\_deg is the difference in angle between the bore elevation and the sight elevation.  Set to undef or -1 to have flight\_simulator() calculate it for you from the zero\_range\_yards parameter (degrees)
>
> parameter: (float) wind\_speed\_fps is the velocity of the wind (feet per second)
>
> parameter: (float) wind\_angle\_deg is the direction the wind is blowing (degrees, 0 = shooting directly into wind, 90 = wind is blowing from the right, perpendicular to flight path, -90 = wind is blowing from the left, perpendicular to flight path)
>
> parameter: (float) OPTIONAL: max\_range\_yards is the maximum range to which the flight will be simulated (yards, default is 2000)
>
> returns: a reference to an array of hash references, one per yard, denoting the projectile's disposition when it reaches that range.  All data fields are floating-point numbers:
>
>     range_yards     How far downrange the projectile has travelled, in yards.
>     drop_inches     How far below the horizontal plane intersecting the muzzle the projectile has travelled, in inches.
>     correction_moa  The angle from the muzzle to the projectile in the vertical plane, relative to the path from the muzzle to the target, in minutes.
>     time_seconds    How much time has elapsed since leaving the muzzle, in seconds.
>     windage_inches  How far in the horizontal plane the projectile has moved due to wind, in inches.
>     windage_moa     The angle from the muzzle to the projectile in the horzontal plane, relative to the path from the muzzle to the target, in minutes.
>     vel_fps         The velocity of the projectile, in feet per second.
>     vel_horiz_fps   The horizontal component of the velocity of the projectile, in feet per second.
>     vel_vert_fps    The vertical component of the velocity of the projectile, in feet per second.

## g1\_drag (velocity\_fps)

The canonical function for computing instantaneous velocity drop at a given velocity, per the G1 drag model.

> parameter: (float) velocity\_fps is the velocity of the projectile (in feet per second)
>
> returns: (float) the deceleration of the projectile from drag (in feet per second per second)

## muzzle\_energy (mass\_grains, velocity\_fps, \[want\_joules\_bool\])

A convenience function for computing kinetic energy from mass and velocity.
Despite its name, it is useful for computing the kinetic energy of a projectile
at any point during its flight.

> parameter: (float) mass\_grains is the mass of the projectile (in grains)
>
> parameter: (float) velocity\_fps is the velocity of the projectile (in feet per second)
>
> parameter: (boolean) OPTIONAL: set want\_joules\_bool to a True value to get Joules instead of foot-pounds (boolean, default=False)
>
> returns: (float) the kinetic energy of the projectile (in foot-pounds or Joules)

## muzzle\_velocity\_from\_energy (mass\_grains, energy\_ftlbs)

A convenience function for computing velocity from mass and kinetic energy.
Despite its name, it is useful for computing the velocity of a projectile
at any point during its flight.

If all you have is Joules, divide by 1.3558179 to get foot-pounds.

> parameter: (float) mass\_grains is the mass of the projectile (in grains)
>
> parameter: (float) energy\_ftlbs is the kinetic energy of the projectile (in foot-pounds)
>
> returns: (float) the velocity of the projectile (in feet per second)

# TODO

Contact Geoffrey Kolbe (Inventor at Border Barrels Limited) and ask for permission to publish his splendid implementations:

    * Bullet drag calculator - http://www.border-barrels.com/cgi-bin/drag_working.cgi

    * Barrel weight calculator - http://www.border-barrels.com/cgi-bin/swamped_barrel_weight.cgi

His drag calculator seems better than my own ebc function.


# NAME

Physics::Ballistics::Terminal -- Terminal ballistics formulae.

# ABSTRACT

Terminal ballistics is the study of what happens when a projectile impacts
its target.  This module implements a variety of functions and mathematical
formulae useful in the analysis and prediction of terminal ballistic effects.

# TWO DOMAINS OF VELOCITY

Some of these functions pertain to the "ballistic" domain, and others pertain
to the "hypervelocity" domain.  These refer to two different and somewhat
ill-defined ranges of velocity.  "Ballistic" velocity ranges from about 300
to 1100 meters per second, while "Hypervelocity" ranges from about 1100 meters
per second to several tens of thousands of meters per second.

Why does this matter?  Because successful models of ballistic interactions
are not accurate in the hypervelocity domain, and successful models of
hypervelocty interactions are not accurate in the ballistic domain.  Thus it
is crucial to use the model which is valid for the velocity of interaction
you are attempting to predict.

Functions which are only valid in one domain or the other will be thus marked.

# REGARDING BULLET DIAMETERS

Some of these functions require the diameter of a projectile as a parameter.
Please note that bullet diameters are usually different from the names of
their calibers.  NATO 5.56mm bullets are actually 5.70mm in diameter, while
Russian 5.45mm bullets are actually 5.62mm.  .308 caliber bullets really are
0.308 inches in diameter (7.82mm), but .22 Long Rifle bullets are 0.222
inches across.  Please do not make assumptions; check before plugging it in!

# DEFINITIONS

## DU

DU is short for "Depleted Uranium".  This denotes any of a number of metallic
alloys containing a high fraction of Uranium, the most common of which is
99.25% Uranium and and 0.75% Titanium.  This material is extremely dense and
hard, and somewhat ductile, making it excellent for armor-piercing projectiles.
Contrary to popular myth, DU projectiles are not "nuclear" and do not explode,
though in the hypervelocity domain they can be pyrophoric.  Nor is DU highly
radioactive.  It is, however, a heavy metal (like lead, mercury, and arsenic)
and therefore toxic.

## MILD STEEL

There are many thousands of steel formulations and treatments, resulting in
myriad different strengths and hardnesses.  "Mild Steel" is a general term
for any of a number of inexpensive steels of moderate to low strength and
hardness.  The steels used in rebar, trench plates, automotive bodies, and
other commodity and construction applications tend to be of this type.  Mild
steels are an inferior material for armor, providing only about 80% of the
protection as an equivalent thickness of RHA.  Mild steel is often used in
military projectile cores (as in America's 7.62mm M80 "Ball") because it is
cheap and much more ductile than hardened steels.  This ductility makes it
less likely to fracture while penetrating its target (whether the target is
made of flesh or harder stuff).

## RHA

RHA is short for "Rolled Homogenous Armor".  It is a commonly-used term for
hardened steel armor material in general, or for armor steel which complies
with the military specification MIL-A-12560.  RHA is also a common standard
material for normalizing depth of penetration.  It is roughly equivalent to
AISI 4340 steel in character, and AISI 4340 is often used in lieu of "real"
RHA in laboratory experiments.

Unless otherwise noted, return values representing a depth of penetration

Changes
0.01     2014-05-04
- Initial port from .calc.code unorganized function repository.

1.00     2017-07-08
- First release.  Finally have it repackaged and suitable for general use.

1.01     2017-07-08
- Fixed Physics::Ballistics package name.

1.02     2017-07-14
- pod2htmd.tmp: Removed turd file.
- lib/Physics/Ballistics/Terminal.pm:  Added clarification to use of momentum in lethality function.
- lib/Physics/Ballistics/Terminal.pm:  Added L<> notation to hyperlinks in poddoc.
- added cpanfile


lib/Physics/Ballistics.pm
package Physics::Ballistics;
use strict;
use warnings;
use vars qw($VERSION $PI);

# Work in progress; currently useless except for documentation.
# Would like this to become an easy, powerful interface akin to http://www.shooterscalculator.com/ballistic-trajectory-chart.php?t=0c84354b
# Generate graphs from multiple profiles via flight_sim, ebc, lethality, etc.
# Include in output statistics about weight and recoil.

use Math::Trig;

BEGIN {
    $VERSION = '1.02';
}

our $PI = Math::Trig::pi;
sub setundef { ${$_[0]} = $_[1] unless ( defined ( ${$_[0]} ) ); }
sub absolute { $_[0] = 0 - $_[0] if ( $_[0] < 0 ); return $_[0]; }

our %FUNCTION_DEPENDENCIES_H = (
  'ebc' => [
      {label => 'proj_mass',  unit => 'grain'},
      {label => 'proj_diam',  unit => 'mm'},
      [{label => 'proj_shape', unit => ''}, {label => 'proj_form_factor', unit => ''}], # list means one or the other must be defined
      ],
  'flight_simulator' => [
      {label => 'drag_model', unit => ''},
      {label => 'proj_bc', unit => ''},
      {label => 'velocity', unit => 'fps'},
      {label => 'sight_height', unit => 'inch'},
      {label => 'shot_angle', unit => 'degree'},
      {label => 'zero_range', unit => 'yard'}
      ],
  'muzzle_energy' => [
      {label => 'proj_mass', unit => 'grain'},
      {label => 'velocity', unit => 'fps'}
      ],
  'muzzle_velocity_from_energy' => [
      {label => 'proj_mass', unit => 'grain'},
      {label => 'proj_ke', unit => 'ftlb'},
      ],
  'cartridge_capacity' => [
      {label => 'proj_diam', unit => 'mm'},
      {label => 'case_base_diam', unit => 'mm'},
      {label => 'case_len', unit => 'mm'},
      {label => 'pressure', unit => 'psi'}
      ],
  'empty_brass' => [
      {label => 'proj_diam', unit => 'mm'},
      {label => 'case_base_diam', unit => 'mm'},
      {label => 'case_len', unit => 'mm'},
      {label => 'pressure', unit => 'psi'}
      ],
  'gunfire' => [
      {label => 'proj_diam', unit => 'mm'},
      {label => 'case_base_diam', unit => 'mm'},
      {label => 'case_len', unit => 'mm'},
      {label => 'pressure', unit => 'psi'},
      {label => 'proj_mass', unit => 'grain'},
      {label => 'barrel_len', unit => 'inch'}
      ],
  'powley' => [
      {label => 'proj_diam', unit => 'mm'},
      {label => 'case_base_diam', unit => 'mm'},
      {label => 'case_len', unit => 'mm'},
      {label => 'barrel_reference_len', unit => 'inch'},
      {label => 'barrel_len', unit => 'inch'}
      ],
  'ke' => [
      {label => 'velocity', unit => 'mps'},
      {label => 'proj_mass', unit => 'grain'}
      ],
  'penetration_rha' => [ # Physics::Ballistics::Terminal::pc()
      {label => 'proj_mass', unit => 'grain'},
      {label => 'velocity', unit => 'fps'},
      {label => 'proj_diam', unit => 'inch'},
      {label => 'proj_composition', unit => ''}
      ],
  'penetration_flesh' => [ # Physics::Ballistics::Terminal::poncelet()
      {label => 'proj_mass', unit => 'grain'},
      {label => 'velocity', unit => 'fps'},
      {label => 'proj_diam', unit => 'inch'},
      {label => 'proj_composition', unit => ''}
      ]
  );

our %FUNCTION_PRODUCT_H = (
  'ebc' => [
      {label => 'proj_bc',  unit => 'grain'},
      ],
  'flight_simulator' => [
      {label => 'flight_ar', unit => 'list'}
      ],
  'muzzle_energy' => [
      {label => 'proj_ke', unit => 'ftlb'}
      ],
  'muzzle_velocity_from_energy' => [
      {label => 'reference_velocity', unit => 'fps'}
      ],
  'cartridge_capacity' => [
      {label => 'case_volume', unit => 'grain_water'}
      ],
  'empty_brass' => [
      {label => 'case_mass', unit => 'grain'}
      ],
  'gunfire' => [
      {label => '', unit => 'hash', mapping => {
          'N*m' => {label => 'proj_ke', unit => 'joule'},
          'f/s' => {label => 'velocity', unit => 'fps'} # zzapp -- whenever setting velocity, also set reference velocity for 24" barrel
          }}
      ],
  'powley' => [
      # special-case -- multiply by reference velocity to get new velocity -- zzapp
      ],
  'ke' => [
      {label => 'proj_ke', unit => 'joule'}
      ],
  'penetration_rha' => [ # Physics::Ballistics::Terminal::pc()
      {label => 'proj_penetration_rha', unit => 'mm'}
      ],
  'penetration_flesh' => [ # Physics::Ballistics::Terminal::poncelet()
      {label => 'proj_penetration_flesh', unit => 'mm'}
      ]
  );

1;

=head1 NAME

Physics::Ballistics -- Ballistics formulae.

=head1 ABSTRACT

Ballistics is the study of the launching, flight and effects of projectiles.
This distribution provides various formulae producing metrics of internal,
external and terminal ballistics.

These formulae are primarily oriented towards the ballistics of small arms 
bullets between 5mm and 14mm and diameter, and to a lesser extent heavier 
hypervelocity projectiles ("long rod" penetrators) and shaped charges.

More extensive documentation is available in each of these three constituent 
modules.

=head1 MODULES

L<Physics::Ballistics::Internal> - launch mechanics

L<Physics::Ballistics::External> - flight mechanics

L<Physics::Ballistics::Terminal> - terminal effects

=head1 TODO

I would like Physics::Ballistics to provide the integrated functionality of 
these modules in an object-oriented way, someday.

Some rocket physics functions are works in progress.  Those might get integrated here eventually.

The units used and returned by these functions are horribly inconsistent, but at least they are documented.  They might be made more consistent in the future.

See also the individual modules for module-specific to-do lists.

=head1 AUTHOR

TTK Ciar, <ttk[at]ciar[dot]org>

=head1 COPYRIGHT AND LICENSE

Copyright 2007-2017 by TTK Ciar

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
